Proposals system. Version 2. Iteration6

shamil-gadelshin opened the pull request:
### Features
- monorepo integration
- introduced `ActorOriginValidator` in the common module for future integration with the membership and council modules
- proposal system authentication system (membership integration) in the engine and discussion modules (`ActorOriginValidator` implementation - `MembershipOriginValidator`)
- council integration for voting: `ActorOriginValidator` and `VotersParameters` implementations with tests - `CouncilManager`
- created tests for `MembershipValidator` and `CouncilManager`
- returned stake after the unstaking (StakingEventsHandler integration)
- removed proposals v1
- runtime integration and implementing all handlers required (council handler, total voters handler)
- change proposals execution model
- move all proposals modules from string error constants to decl_error! approach
- jammed 'srml_staking_reward_curve::build!' [warning](https://github.com/Joystream/substrate-runtime-joystream/issues/179)
- `StakingEventsHandler` integration with the `ContentWorkingGroupStakingEventHandler`
- `set_countil()` is public now for testing purposes (it's guarded by `ensure_root()`))

### Related issues

- https://github.com/Joystream/substrate-runtime-joystream/issues/174
- https://github.com/Joystream/substrate-runtime-joystream/issues/179
- https://github.com/Joystream/substrate-runtime-joystream/issues/154

### [Iteration 5 features (was skipped)](https://github.com/Joystream/substrate-runtime-joystream/pull/176)
- spam gating: prevent the same author from writing more than N posts in a row.
- change `EnsureOrigin` to the `ActorOriginValidator` model (was rewritten during the 6th iteration )
- create events for the discussion system

shamil-gadelshin committed:
Add antispam-gate for the thread creation

Add ‘max number of threads by same author in a row’ limit

shamil-gadelshin committed:
Update comments

shamil-gadelshin committed:
Add ActorOriginValidator trait

- add ActorOriginValidator trait
- update create_thread() with ThreadAuthorOriginValidator

shamil-gadelshin committed:
Introduce ‘PostAuthorOriginValidator’ associated type

- upgrade add_post() and update_post() with PostAuthorOriginValidator

shamil-gadelshin committed:
Add events to the discussion module

- add ThreadCreated, PostCreated, PostUpdated events
- add tests
- move error messages to the separate module

shamil-gadelshin committed:
Update codex module with latest discussion module

- add ThreadAuthorId param stub
- update mock and tests

shamil-gadelshin committed:
Clean content-working-group tests

shamil-gadelshin committed:
Update obsolete code in the runtime/build.rs

Migrated to the WasmBuilder.

mnaamani committed:
Merge pull request #193 from shamil-gadelshin/runtime_warning_cleanup

Runtime warning cleanup

shamil-gadelshin committed:
Merge branch 'proposals_v2' into proposals_v2_iteration6

shamil-gadelshin committed:
Move proposals to the ‘runtime-modules’ folder

shamil-gadelshin committed:
Update Cargo.toml’s

- delete proposals workspace
- update proposals Cargo.toml’s (engine, codex, discussions)
- update global workspace

shamil-gadelshin committed:
Merge branch 'development' into proposals_v2_iteration6

shamil-gadelshin committed:
Disable warning in runtime lib.rs

Add #![allow(array_into_iter)] to runtime/lib.rs

srml_staking_reward_curve::build! - substrate macro produces a warning.
We need to remove this attribute after post-Rome substrate upgrade.

shamil-gadelshin committed:
Integrate discussion system (proposals) with membership crate

shamil-gadelshin committed:
Convert PostAuthorId to the MemberId

- convert PostAuthorId to the MemberId in proposal discussion module
- change validate_actor_origin() to the ensure_actor_origin()
- convert ensure_actor_origin() from bool to Result<(), &’static str>

shamil-gadelshin committed:
Integrate engine with membership module

shamil-gadelshin committed:
Refactor actor_origin_validator

- move actor_origin_validator trait and implementation into the membership crate
- add actor_validation to the engine

shamil-gadelshin committed:
Update & fix tests

shamil-gadelshin committed:
Add tests for membership module (MembershipOriginValidator)

shamil-gadelshin committed:
Update membership module tests (origin validator)

shamil-gadelshin committed:
Make set_countil() public for testing purposes

shamil-gadelshin committed:
Move ActorOriginValidator from membership to common

shamil-gadelshin committed:
Integrate engine module with council

- add governance dependency
- implement ActorOriginValidator for council voting (CouncilOriginValidator)
- change mocks
- add tests

shamil-gadelshin committed:
Implement total_voters_count() for engine module

- implement VotersParameters::total_voters_count() using council size for proposals engine module
- cargo fmt

shamil-gadelshin committed:
Change ensure_actor_origin() signature

- change ensure_actor_origin() signature - remove error
- introduce built-in error messages

shamil-gadelshin committed:
Add souce_account_id in the engine module

- create new StakeData structure with stake_id field in the engine module
- add souce_account_id field to the stake_data
- change stake_id field to the StakeData in the Proposal structure

shamil-gadelshin committed:
Add StakingEventHandler for engine module

- add StakingEventHandler for engine module
- add storage map StakesProposals
- add refund_proposal_stake() callback

shamil-gadelshin committed:
Remove comments

shamil-gadelshin committed:
Remove proposals version 1

- delete proposals from the governance module
- clear the runtime from link to proposals version 1

shamil-gadelshin committed:
Add proposals modules to the runtime

shamil-gadelshin committed:
Fix StakeEventsHandler in the proposals engine

shamil-gadelshin committed:
Integrate proposals StakeEventsHandler

- integrate proposals StakeEventsHandler with the ContentWorkingGroupStakingEventHandler

shamil-gadelshin committed:
Add proposals extrinsics to the proposals codex

- add proposals extrinsics to the proposals codex to start conversion from ProposalTypes to proposal extrinsics execution model

shamil-gadelshin committed:
Migrate to the extrinsic execution model

shamil-gadelshin committed:
Restore tests for the extrinsic execution model

- delete custom proposal types
- create test extrinsic
- change ProposalCode type

shamil-gadelshin committed:
Migrate proposals engine to decl_error! macro

- introduce errors in decl_error! macro for the crate
- delete errors module
- fix tests

shamil-gadelshin committed:
Migrate proposals to decl_error! macro

- migrate discussions
- create codex errors conversion helpers
- update tests

shamil-gadelshin committed:
Apply cargo fmt

bedeho reviewed (changes requested):
- This review goes beyond the diff in the PR, because it was frankly really hard to just follow that, given how long ago I looked at this, and also because prior reviews did not go that in depth on different extrinsics, as we were just getting started.

- In general, types with postfix "data" indicate a sloppy name or bad type, unless they literally represent unstructured data without any internal meaning. There is at least `StakeData` and `FinalizationData` in this iteration. For the former, I think it should not exist (as you will see in the review), and for the latter, see next section.

- Question: Is there any reason to check origin on pure `Module` methods? doesnet seem to make sense? the calling code will perform the origin check. Perhaps this is why we have the akward double origin thing... its not needed?

# Engine

- When a council changed, I think we agreed to reset all the current votes and reset the finalization deadline. That is currently not implemented, right?

- `execute_proposal` appears to have a bug, as it unconditionally removes from `PendingExecutionProposalIds`, even if the first guard rejects the proposal. Moreover, the routine does not check that the proposal actually exists, yet it does check that it has the right state. This seems a bit inconsistent. I think that if its going to make an assumption without checking, it needs not check for any even weaker assumptions. Moreover, its also a bit akward that it is rechecking whether the given proposal is finalized, despite the caller already having done this check in computation `get_approved_proposal_with_expired_grace_period_ids`, that should be avoidable in the same code path by picking the right representations. Have a look, it may not be worth changing so much, but I wanted to draw your attention to it.

- For `FinalizationData`
  - I think there is a crisper name could be `FinalizedProposalStatus`
  - The field `finalization_error` should reflect that this is specifically the serialized error associated with unstaking. When reading it I first thought this is what held the error related to execution as well, but that is indeed separate, so name should be sharpened. Also, something needs to be done to very clearly signal that this is no normal error, but indeed the broken invariant error. At the very least the comment must say so, but perhaps also the name, e.g. `encoded_unstaking_error_due_to_broken_runtime`. Its a bit long, but the length helps to grab the attention to how distinct this field is.
  - Lets drop `finalized_at`. Its purely archival, and so query node should hold it, and it's not even clear that we will, in the end, hold the entire proposal itself after it is executed.

- Did we not decide that, for the time being, we will not store broken invariant errors in storage, simply for the benefit of archiving error messages. We would use either console logging or something else to capture them? Unfortunately, we should have written down what we decided. At any rate, this sort of storing of errors seems to do little if anything of increasing safety or blocking further state corruption, which are the strong reasons for handling broken runtimes in some way. 

- The only reason we have `ActiveProposalIds` is so that more effectively iterate only the proposals that are active when trying to finalize. This also why we need to limit how many can be active at once.

  This is as intended, but I want to explore an alternative. If you think its an improvement, lets make an issue out of it and have someone work on it at a later time:

  We drop `ActiveProposalIds` and the limit on the number of active proposals, as well as any iteration in `on_finalized`. Instead, a new map `finalize_at_block: T::BlockNumber => T::ProposalId` is introduced, and a mapping is introduced whenever a proposal is created. This map is checked every `on_finalized`, but there is no iteration, you just check for the specific block you are on. There is an implicit limit that only one proposal can be finalized per block, and that is enforced upon proposal creation. 

  It seems this is both safer, as there is no enumeration and iteration, and we drop one parameter value, and we have no limit on the number of simultaneous proposals in flight.

  What do you think?

  The same would apply to `PendingExecutionProposalIds`.

- Seems like this storage map should be dropped, and instead just hold the value as a field in the `struct Proposal`.

  `pub ProposalCode get(fn proposal_codes): map T::ProposalId =>  Vec<u8>;`

# Codex

Looks great!

# Discussion system

Looks great!

shamil-gadelshin committed:
Move origin validators to the runtime

Move CouncilManager, MemberhipOriginValidator to the runtime.

Issue link: https://github.com/Joystream/substrate-runtime-joystream/issues/226

shamil-gadelshin committed:
Fix integration tests

Fix tests for moved CouncilManager and MemberhipOriginValidator in the runtime proposal integration

shamil-gadelshin committed:
Move StakingEventsHandler to the runtime

- move StakingEventsHandler to the runtime integration folder from the proposals engine module.

shamil-gadelshin committed:
Refactor proposals engine module

- rename StakeData to the ActiveStake
- rename ProposalCode to the DispatchableCallCode
- rename ProposalObject to the ProposalOf

shamil-gadelshin committed:
Add ensure calls of dependent modules in codex

Add calls:
- add ensure_create_proposal_parameters_are_valid() call
- add ensure_can_create_thread() call

to the create_proposals_* extrinsics in the codex

shamil-gadelshin committed:
Change proposals actor validation

- move memberhip validation to the codex proposal creation extrinsics
- remove actor validation in the create_proposal (engine) and create_thread(discussion) API methods.

shamil-gadelshin committed:
Change MembershipOriginValidator in the runtime

- remove root_account_id from valid member origin

shamil-gadelshin committed:
Refactor execute_proposal() in the engine module

shamil-gadelshin committed:
Refactor proposal status finalization_error

- rename it to the encoded_unstaking_error_due_to_broken_runtime

shamil-gadelshin committed:
Refactor proposal stakes

- move proposal stake data from the proposal to its statuses

bedeho reviewed (approved):
Great job!
